Documentation changes for new tools modifications for ARMC5/ARMC6 usage (#1024)
ARMmbed/Handbook: docs/reference/configuration/mbed_targets.md

The `supported_toolchains` property is the list of toolchains that support a target. The allowed values for `supported_toolchains` are `ARM`, `uARM`, `ARMC5`, `ARMC6`, `GCC_ARM` and `IAR`.

When using `ARM`, `ARMC5`, `ARMC6` for `supported_toolchains`, please note:

> - If you specify `ARMC5` in `supported_toolchains`, it means the target supports Arm Compiler 5 (version 5.06 update 6) only, and the build system uses Arm Compiler 5 when you specify `-t ARM` or `-t ARMC5` with `mbed compile` command. 

Reviewer: If the target supports Arm Compiler 5, will `supported_toolchains` show `ARMC5`, or do you have to manually add that?

Author: @AnotherButler - Good question, the changes which Brian is working on will add everything as required and according to these rules. So no manual addition required from user or the reader of this documentation. But I said "If you specify" because this document is talking about "Adding and configuring targets".

Reviewer: So if I changed it to "If the target supports Arm Compiler 5 (version 5.06 update 6) only, `supported_toolchains` will specify `ARMC5`", would that be correct?

Author: @AnotherButler - It is technically correct. But I think the intended audience for this document is someone who is adding or configuring a target, as in the document title. That's why its written as "if you specify.." meaning whoever is adding or configuring a target. But I also want that information to be available for anyone, so I created a link from cli-compile.md to here. So, Ill leave that up to you and if the new format is cleaner/compact I would keep that.

- If the target supports compiling with both Arm Compiler 5 and Arm Compiler 6, `supported_toolchains` specifies `ARM`. Arm Compiler 6 is the default Arm Compiler for Mbed OS. If a target specifies `ARM` in `supported_toolchains`, the system defaults to using Arm Compiler 6 when you use `-t ARM` with the `mbed compile` command. If the build system fails to detect a valid configuration for Arm Compiler 6, it automatically (without any manual intervention) detects if a valid configuration for Arm Compiler 5 exists and continues using Arm Compiler 5 if it successfully detects Arm Compiler 5. You will see a warning message in your standard output from the build system indicating this behavior.

- You can only specify one of `ARM`, `ARMC5` or `ARMC6` for `supported_toolchains` for a target in `targets/targets.json` or `custom_targets.json`.

- It's invalid to use `-t ARMC5` with the `mbed compile` command for a target that specifies `ARMC6` for `supported_toolchains`, and it's invalid to use `-t ARMC6` if `supported_toolchains` specifies `ARMC5`.

> <span class="notes">**Note:** Arm Compiler 6 is the default Arm Compiler for Mbed OS development. Most Mbed OS platforms are already compatible with Arm Compiler 6. Some existing targets still supporting Arm Compiler 5 will be migrated to Arm Compiler 6 in the future. Please be aware that you must use Arm Compiler 6 for future development because Arm Compiler 5 support will be deprecated in September 2019.</span>

Reviewer 2: Can we add a note here saying that it's free for Mbed usecases and link to instructions how the get it?

Reviewer 3: I can do that - do you happen to know where those instructions are? I'm not familiar with this bit.

Reviewer 3: Ah, found something: https://developer.arm.com/products/software-development-tools/compilers/arm-compiler/downloads/version-6

> <span class="notes">**Note:** Arm Compiler 6 is the default Arm Compiler for Mbed OS development. Most Mbed OS platforms are already compatible with Arm Compiler 6. Some existing targets still supporting Arm Compiler 5 will be migrated to Arm Compiler 6 in the future. Please be aware that you must use Arm Compiler 6 for future development because Arm Compiler 5 support will be deprecated in September 2019.</span>

Reviewer 3: ```suggestion
<span class="notes">**Note**: Arm Compiler 6 is the default Arm Compiler version for Mbed OS development, and is [free for use with Mbed OS]((https://os.mbed.com/docs/mbed-os/development/tools/index.html#compiler-versions). Most platforms are already compatible with it; platforms still supporting Arm Compiler 5 will be migrated to Arm Compiler 6. Please do not use Arm Compiler 5 in any new development, as its support will be deprecated in September 2019.
.</span>
```

Reviewer 3: @SenRamakri this is a fork, not a branch. I therefore don't have permission to push these suggestions. Please push if you approve them.

- If the target supports compiling with both Arm Compiler 5 and Arm Compiler 6, `supported_toolchains` specifies `ARM`. Arm Compiler 6 is the default Arm Compiler for Mbed OS. If a target specifies `ARM` in `supported_toolchains`, the system defaults to using Arm Compiler 6 when you use `-t ARM` with the `mbed compile` command. If the build system fails to detect a valid configuration for Arm Compiler 6, it automatically (without any manual intervention) detects if a valid configuration for Arm Compiler 5 exists and continues using Arm Compiler 5 if it successfully detects Arm Compiler 5. You will see a warning message in your standard output from the build system indicating this behavior.

- You can only specify one of `ARM`, `ARMC5` or `ARMC6` for `supported_toolchains` for a target in `targets/targets.json` or `custom_targets.json`.

- It's invalid to use `-t ARMC5` with the `mbed compile` command for a target that specifies `ARMC6` for `supported_toolchains`, and it's invalid to use `-t ARMC6` if `supported_toolchains` specifies `ARMC5`.

> <span class="notes">**Note**: Arm Compiler 6 is the default Arm Compiler version for Mbed OS development, and is [free for use with Mbed OS]((https://os.mbed.com/docs/mbed-os/development/tools/index.html#compiler-versions). Most platforms are already compatible with it; platforms still supporting Arm Compiler 5 will be migrated to Arm Compiler 6. Please do not use Arm Compiler 5 in any new development, as its support will be deprecated in September 2019.

Reviewer 3: My mistake, in both notes, there are two `((` in the link; there should be only one

The `supported_toolchains` property is the list of toolchains that support a target. The allowed values for `supported_toolchains` are `ARM`, `uARM`, `ARMC5`, `ARMC6`, `GCC_ARM` and `IAR`.

There are few rules and tools behavior you should be aware of when using `ARM`, `ARMC5`, `ARMC6` for `supported_toolchains` as follows.
* If you specify `ARMC5` in `supported_toolchains`, it means the target supports `Arm Compiler 5 (version 5.06 update 6)` only and the build system will use `Arm Compiler 5` when you specify `-t ARM` or `-t ARMC5` with `mbed compile` command. 

* If you specify `ARMC6` in `supported_toolchains`, it means the target supports `Arm Compiler 6 (version 6.11)` only and the build system will use `Arm Compiler 6` when you specify `-t ARM` or `-t ARMC6` with `mbed compile` command.

* If a target specify `ARM` in `supported_toolchains`, it means the target supports compiling with both `Arm Compiler 5` and `Arm Compiler 6`. Note that starting with `MbedOS 5.12 release`, `Arm Compiler 6` will 

> be the default ARM Compiler for MbedOS. So, for a target which specifies `ARM` in `supported_toolchains`, the system will default to using `Arm Compiler 6` when you specify `-t ARM` with `mbed compile` command. If the build system

Reviewer 4: Missing a space here:
```suggestion
be the default ARM Compiler for Mbed OS. So, for a target which specifies `ARM` in `supported_toolchains`, the system will default to using `Arm Compiler 6` when you specify `-t ARM` with `mbed compile` command. If the build system
```

> fails to detect a valid configuration for `Arm Compiler 6`, it will automatically(without any manual intervention) try to detect if a valid confguration for `ARM Compiler 5` exists, and will continue using `ARM Compiler 5` if it successfully detects `ARM Compiler 5`. You may see a warning message in your standard output from build system indicating this behavior. 

Reviewer 4: I suggest being more definite when documenting the behavior of the tools:
```suggestion
fails to detect a valid configuration for `Arm Compiler 6`, it will automatically(without any manual intervention) try to detect if a valid configuration for `ARM Compiler 5` exists, and will continue using `ARM Compiler 5` if it successfully detects `ARM Compiler 5`. You will see a warning message in your standard output from the build system indicating this behavior. 
```

> * You can only specify one of `ARM`, `ARMC5` or `ARMC6` for `supported_toolchains` for a target in `targets/targets.json` or `custom_targets.json`.

Reviewer 4: Good to know, I'm not currently erroring on this 😄. I can make this change.
